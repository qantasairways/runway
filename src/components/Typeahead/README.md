```js
<Typeahead
  label="From"
  items={[
    { text: 'Sydney, Australia', badge: 'SYD' },
    { text: 'Melbourne, Australia', badge: 'MEL' },
    { text: 'Brisbane, Australia', badge: 'BNE' },
    { text: 'Perth, Australia', badge: 'PER' },
    { text: 'Darwin, Australia', badge: 'DRW' },
    { text: 'Gold Coast, Australia', badge: 'OOL' },
    { text: 'Adelaide, Australia', badge: 'ADL' },
    { text: 'Los Angeles, United States', badge: 'LAX' },
    { text: 'Abu Dhabi, United Arab Emirates', badge: 'AUH' },
    { text: 'Toowoomba (Brisbane West Wellcamp), Australia', badge: 'WTB' },
    { text: 'London (Heathrow), United Kingdom', badge: 'LHR' },
    { text: 'New York (Newark), United States', badge: 'EWR' }
  ]}
  itemToString={item => (item ? String(item.text) : '')}
  badgeToString={item => (item ? String(item.badge) : '')}
  menuHeight="auto"
  filterItems={(items, inputValue) => {
    return items.filter(item => {
      const regex = new RegExp(inputValue, 'gi');

      return regex.test(`${item.text} ${item.badge}`);
    });
  }}
/>
```

### When selectedItemValue provided

Optional prop to maintain state externally (passed to downshift's selectedItem prop)

```js
<StatefulManager initial={{ text: 'Gold Coast, Australia', badge: 'OOL' }}>
  {({ value, updater }) => (
    <Typeahead
      label="From"
      items={[
        { text: 'Sydney, Australia', badge: 'SYD' },
        { text: 'Melbourne, Australia', badge: 'MEL' },
        { text: 'Brisbane, Australia', badge: 'BNE' },
        { text: 'Perth, Australia', badge: 'PER' },
        { text: 'Darwin, Australia', badge: 'DRW' },
        { text: 'Gold Coast, Australia', badge: 'OOL' },
        { text: 'Adelaide, Australia', badge: 'ADL' },
        { text: 'Los Angeles, United States', badge: 'LAX' },
        { text: 'Abu Dhabi, United Arab Emirates', badge: 'AUH' },
        { text: 'Toowoomba (Brisbane West Wellcamp), Australia', badge: 'WTB' },
        { text: 'London (Heathrow), United Kingdom', badge: 'LHR' },
        { text: 'New York (Newark), United States', badge: 'EWR' }
      ]}
      itemToString={item => (item ? String(item.text) : '')}
      badgeToString={item => (item ? String(item.badge) : '')}
      menuHeight="auto"
      filterItems={(items, inputValue) => {
        return items.filter(item => {
          const regex = new RegExp(inputValue, 'gi');

          return regex.test(`${item.text} ${item.badge}`);
        });
      }}
      selectedItemValue={value}
      onChange={updater}
    />
  )}
</StatefulManager>
```
