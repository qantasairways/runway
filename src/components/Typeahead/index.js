/* eslint jsx-a11y/label-has-associated-control: "off",
jsx-a11y/label-has-for: "off" */ // These attributes are generated by the Downshift library

import React, { Component } from 'react';
import Downshift from 'downshift';
import PropTypes from 'prop-types';

import { colours } from '../../theme/airways';

class Typeahead extends Component {
  onInputValueChange = (value, stateAndHelpers) => {
    const { fetchListOnInput } = this.props;
    if (
      fetchListOnInput &&
      stateAndHelpers.type === Downshift.stateChangeTypes.changeInput
    ) {
      fetchListOnInput(value);
    }
  };

  filterItems = (items, inputValue) => {
    const { itemToString, filterItems } = this.props;
    return filterItems
      ? filterItems(items, inputValue)
      : items.filter(item => {
          const regex = new RegExp(inputValue, 'gi');
          return regex.test(itemToString(item));
        });
  };

  renderItems = (
    getMenuProps,
    getItemProps,
    highlightedIndex,
    selectedItem,
    inputValue
  ) => {
    const { items, itemToString, fetchListOnInput } = this.props;
    const filteredItems = fetchListOnInput
      ? items
      : this.filterItems(items, inputValue);

    return (
      <ul {...getMenuProps()}>
        {filteredItems.map((item, index) => (
          <li
            key={itemToString(item)}
            {...getItemProps({
              index,
              item
            })}
            css={{
              backgroundColor:
                highlightedIndex === index
                  ? colours.highlights
                  : colours.transparent,
              borderBottom: '1px solid lightgrey',
              fontWeight: selectedItem === item ? 'bold' : 'normal',
              listStyleType: 'none',
              padding: '5px'
            }}
          >
            {itemToString(item)}
          </li>
        ))}
      </ul>
    );
  };

  render() {
    const {
      className,
      disabled,
      id,
      initialSelectedItem,
      isFetchingList,
      itemToString,
      label,
      minChars,
      onBlur,
      onChange,
      message,
      placeholder,
      stateReducer,
      valid,
      selectItemCollector,
      maxLength
    } = this.props;

    return (
      <Downshift
        initialSelectedItem={initialSelectedItem}
        isFetchingList={isFetchingList}
        itemToString={itemToString}
        onChange={onChange}
        onInputValueChange={this.onInputValueChange}
        stateReducer={stateReducer}
      >
        {({
          getInputProps,
          getItemProps,
          getLabelProps,
          getMenuProps,
          isOpen,
          inputValue,
          highlightedIndex,
          selectedItem,
          selectItem
        }) => {
          if (selectItemCollector) {
            selectItemCollector(selectItem);
          }
          return (
            <div>
              {label && <label {...getLabelProps()}>{label}</label>}
              <div
                css={{ width: '100%', maxWidth: '500px' }}
                className={className}
              >
                <input
                  css={{
                    padding: '5px',
                    boxSizing: 'border-box',
                    '&:focus': {
                      outlineColor: colours.highlights
                    },
                    width: '100%'
                  }}
                  {...getInputProps({
                    disabled,
                    id,
                    placeholder,
                    onBlur,
                    maxLength
                  })}
                />
                <ul {...getMenuProps()}>
                  {isOpen && !isFetchingList && inputValue.length >= minChars
                    ? this.renderItems(
                        getMenuProps,
                        getItemProps,
                        highlightedIndex,
                        selectedItem,
                        inputValue
                      )
                    : null}
                  {isFetchingList && <span>Loading...</span>}
                </ul>
              </div>
              {message && !valid && <div>{message}</div>}
            </div>
          );
        }}
      </Downshift>
    );
  }
}

Typeahead.propTypes = {
  className: PropTypes.string,
  disabled: PropTypes.bool,
  fetchListOnInput: PropTypes.func,
  id: PropTypes.string,
  initialSelectedItem: PropTypes.shape(),
  isFetchingList: PropTypes.bool,
  items: PropTypes.arrayOf(PropTypes.shape),
  itemToString: PropTypes.func,
  label: PropTypes.string,
  onBlur: PropTypes.func,
  onChange: PropTypes.func,
  onInputValueChange: PropTypes.func,
  filterItems: PropTypes.func,
  message: PropTypes.string,
  minChars: PropTypes.number,
  placeholder: PropTypes.string,
  stateReducer: PropTypes.func,
  valid: PropTypes.bool,
  selectItemCollector: PropTypes.func,
  maxLength: PropTypes.number
};

Typeahead.defaultProps = {
  ...Downshift.propTypes,
  className: null,
  disabled: false,
  fetchListOnInput: undefined,
  id: null,
  initialSelectedItem: null,
  isFetchingList: false,
  itemToString: item => (item ? String(item) : ''),
  items: [],
  label: '',
  onBlur: undefined,
  onChange: undefined,
  onInputValueChange: undefined,
  message: '',
  minChars: 0,
  placeholder: '',
  stateReducer: undefined,
  valid: true,
  maxLength: 100
};

export default Typeahead;
