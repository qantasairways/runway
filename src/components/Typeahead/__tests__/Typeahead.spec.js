import React, { Component } from 'react';
import { mount } from 'enzyme';
import Downshift from 'downshift';

import Typeahead from '..';

const items = ['apple', 'pear', 'orange', 'grape', 'banana'];
const itemsCustom = [
  { value: 'apple' },
  { value: 'pear' },
  { value: 'orange' },
  { value: 'grape' },
  { value: 'banana' }
];

describe('Typeahead', () => {
  it('renders correctly with defaults', () => {
    const component = mount(<Typeahead items={[]} />);

    expect(component).toMatchSnapshot();
  });

  it('renders correctly with items', () => {
    const component = mount(<Typeahead items={items} />);

    expect(component).toMatchSnapshot();
  });

  it('renders correctly with error state', () => {
    const component = mount(
      <Typeahead items={items} valid={false} message="Something went wrong" />
    );

    expect(component).toMatchSnapshot();
  });

  it('renders correctly when fetching', () => {
    const component = mount(<Typeahead items={items} isFetchingList />);
    expect(component).toMatchSnapshot();
  });

  describe('filterItems()', () => {
    it('returns the matching items for simple items array', () => {
      const component = mount(<Typeahead items={items} />);
      expect(component.instance().filterItems(items, 'ra')).toEqual([
        'orange',
        'grape'
      ]);
    });

    it('returns the matching items for custom structure of items', () => {
      const itemToString = item => (item ? String(item.value) : '');

      const component = mount(
        <Typeahead items={itemsCustom} itemToString={itemToString} />
      );
      expect(component.instance().filterItems(itemsCustom, 'ra')).toEqual([
        { value: 'orange' },
        { value: 'grape' }
      ]);
    });

    it('calls the filter prop method if provided', () => {});
  });

  describe('renderItems()', () => {
    it('returns the correct jsx for simple items array', () => {
      const component = mount(<Typeahead items={items} />);
      expect(
        component.instance().renderItems(() => {}, () => {}, 5, 6, 'ra')
      ).toMatchSnapshot();
    });

    it('returns the correct jsx for custom structure of items', () => {
      const itemToString = item => (item ? String(item.value) : '');

      const component = mount(
        <Typeahead items={itemsCustom} itemToString={itemToString} />
      );
      expect(
        component.instance().renderItems(() => {}, () => {}, 5, 6, 'ra')
      ).toMatchSnapshot();
    });
  });

  describe('onInputValueChange()', () => {
    let fetchListOnInputMock = jest.fn();

    beforeEach(() => {
      fetchListOnInputMock = jest.fn().mockResolvedValue({});
    });

    afterEach(() => {
      fetchListOnInputMock.mockReset();
    });

    it('does nothing when fetchListOnInput prop does not exist', () => {
      const component = mount(<Typeahead items={items} />);
      component.instance().onInputValueChange('grape', { selectedItem: null });
      expect(fetchListOnInputMock.mock.calls.length).toBe(0);
    });

    it('does not call fetchListOnInput() when input change type is not "changeInput"', () => {
      const component = mount(
        <Typeahead items={items} fetchListOnInput={fetchListOnInputMock} />
      );
      const downshiftChangetype = Downshift.stateChangeTypes.clickItem;
      component.instance().onInputValueChange('grape', {
        selectedItem: null,
        type: downshiftChangetype
      });
      expect(fetchListOnInputMock.mock.calls.length).toBe(0);
    });

    it('calls fetchListOnInput() when input change type is "changeInput"', () => {
      const downshiftChangetype = Downshift.stateChangeTypes.changeInput;
      const component = mount(
        <Typeahead
          items={items}
          minChars={3}
          fetchListOnInput={fetchListOnInputMock}
        />
      );
      component.instance().onInputValueChange('grap', {
        selectedItem: null,
        type: downshiftChangetype
      });
      expect(fetchListOnInputMock.mock.calls.length).toBe(1);
    });
  });

  describe('Escape Hatches', () => {
    let fetchListOnInputMock = jest.fn();
    beforeEach(() => {
      fetchListOnInputMock = jest.fn().mockResolvedValue({});
    });
    afterEach(() => {
      fetchListOnInputMock.mockReset();
    });
    describe('selectItemCollector', () => {
      it('should hand selectItem to selectItemCollector and allow selected state overrides', () => {
        // naive implementation of using escape hatch to override Typeaheads internal
        // state selection for testing purposes.
        class Tracker extends Component {
          track = collectedFx => {
            this.tracked = collectedFx;
          };

          render = () => (
            <Typeahead
              items={items}
              minChars={3}
              fetchListOnInput={fetchListOnInputMock}
              selectItemCollector={(...args) => this.track(...args)}
            />
          );
        }

        const tracker = mount(<Tracker />);

        const { value: initialSelection } = tracker.find('input').props();
        expect(initialSelection).toEqual('');

        const desiredSelection = 'mock desired new selection';
        tracker.instance().tracked(desiredSelection);
        tracker.update();

        const { value: updatedSelection } = tracker.find('input').props();
        expect(updatedSelection).toEqual(desiredSelection);
      });
    });
  });

  describe('selectedItemValue()', () => {
    let fetchListOnInputMock = jest.fn();

    beforeEach(() => {
      fetchListOnInputMock = jest.fn().mockResolvedValue({});
    });

    afterEach(() => {
      fetchListOnInputMock.mockReset();
    });

    it('should handle selectedItemValue to selectedItem and allow selectedItem state externally', () => {
      const component = mount(
        <Typeahead
          items={items}
          minChars={3}
          fetchListOnInput={fetchListOnInputMock}
          selectedItemValue={items[2]}
        />
      );
      const { value: initialSelection } = component.find('input').props();
      expect(initialSelection).toEqual('orange');
    });

    it('should handle selectedItemValue when value is empty', () => {
      const component = mount(
        <Typeahead
          items={items}
          minChars={3}
          fetchListOnInput={fetchListOnInputMock}
          selectedItemValue=""
        />
      );
      const { value: initialSelection } = component.find('input').props();
      expect(initialSelection).toEqual('');
    });
  });
});
