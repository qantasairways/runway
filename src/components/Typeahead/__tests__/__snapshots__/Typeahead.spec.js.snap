// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Typeahead renderItems() returns the correct jsx for custom structure of items 1`] = `
<ul>
  <li
    className="css-1l18fj6"
  >
    orange
  </li>
  <li
    className="css-1l18fj6"
  >
    grape
  </li>
</ul>
`;

exports[`Typeahead renderItems() returns the correct jsx for simple items array 1`] = `
<ul>
  <li
    className="css-1l18fj6"
  >
    orange
  </li>
  <li
    className="css-1l18fj6"
  >
    grape
  </li>
</ul>
`;

exports[`Typeahead renders correctly when fetching 1`] = `
.emotion-1 {
  width: 100%;
  max-width: 500px;
}

.emotion-0 {
  padding: 5px;
  box-sizing: border-box;
  width: 100%;
}

.emotion-0:focus {
  outline-color: #8de2e0;
}

<Typeahead
  className={null}
  defaultHighlightedIndex={[Function]}
  defaultIsOpen={[Function]}
  disabled={false}
  environment={[Function]}
  getA11yStatusMessage={[Function]}
  getItemId={[Function]}
  highlightedIndex={[Function]}
  id={null}
  initialHighlightedIndex={[Function]}
  initialInputValue={[Function]}
  initialIsOpen={[Function]}
  initialSelectedItem={null}
  inputId={[Function]}
  inputValue={[Function]}
  isFetchingList={true}
  isOpen={[Function]}
  itemCount={[Function]}
  itemToString={[Function]}
  items={
    Array [
      "apple",
      "pear",
      "orange",
      "grape",
      "banana",
    ]
  }
  label=""
  labelId={[Function]}
  maxLength={100}
  menuId={[Function]}
  message=""
  minChars={0}
  onOuterClick={[Function]}
  onSelect={[Function]}
  onStateChange={[Function]}
  onUserAction={[Function]}
  placeholder=""
  scrollIntoView={[Function]}
  selectedItem={[Function]}
  selectedItemChanged={[Function]}
  suppressRefError={[Function]}
  valid={true}
>
  <Downshift
    defaultHighlightedIndex={null}
    defaultIsOpen={false}
    environment={[Window]}
    getA11yStatusMessage={[Function]}
    initialSelectedItem={null}
    isFetchingList={true}
    itemToString={[Function]}
    onChange={[Function]}
    onInputValueChange={[Function]}
    onOuterClick={[Function]}
    onSelect={[Function]}
    onStateChange={[Function]}
    onUserAction={[Function]}
    scrollIntoView={[Function]}
    selectedItemChanged={[Function]}
    stateReducer={[Function]}
    suppressRefError={false}
  >
    <div
      aria-expanded={false}
      aria-haspopup="listbox"
      aria-labelledby="downshift-3-label"
      aria-owns={null}
      role="combobox"
    >
      <div
        className="emotion-1 null"
      >
        <input
          aria-activedescendant={null}
          aria-autocomplete="list"
          aria-controls={null}
          aria-labelledby="downshift-3-label"
          autoComplete="off"
          className="emotion-0"
          disabled={false}
          id={null}
          maxLength={100}
          onBlur={[Function]}
          onChange={[Function]}
          onKeyDown={[Function]}
          placeholder=""
          value=""
        />
        <ul
          aria-labelledby="downshift-3-label"
          id="downshift-3-menu"
          role="listbox"
        >
          <span>
            Loading...
          </span>
        </ul>
      </div>
    </div>
  </Downshift>
</Typeahead>
`;

exports[`Typeahead renders correctly with defaults 1`] = `
.emotion-1 {
  width: 100%;
  max-width: 500px;
}

.emotion-0 {
  padding: 5px;
  box-sizing: border-box;
  width: 100%;
}

.emotion-0:focus {
  outline-color: #8de2e0;
}

<Typeahead
  className={null}
  defaultHighlightedIndex={[Function]}
  defaultIsOpen={[Function]}
  disabled={false}
  environment={[Function]}
  getA11yStatusMessage={[Function]}
  getItemId={[Function]}
  highlightedIndex={[Function]}
  id={null}
  initialHighlightedIndex={[Function]}
  initialInputValue={[Function]}
  initialIsOpen={[Function]}
  initialSelectedItem={null}
  inputId={[Function]}
  inputValue={[Function]}
  isFetchingList={false}
  isOpen={[Function]}
  itemCount={[Function]}
  itemToString={[Function]}
  items={Array []}
  label=""
  labelId={[Function]}
  maxLength={100}
  menuId={[Function]}
  message=""
  minChars={0}
  onOuterClick={[Function]}
  onSelect={[Function]}
  onStateChange={[Function]}
  onUserAction={[Function]}
  placeholder=""
  scrollIntoView={[Function]}
  selectedItem={[Function]}
  selectedItemChanged={[Function]}
  suppressRefError={[Function]}
  valid={true}
>
  <Downshift
    defaultHighlightedIndex={null}
    defaultIsOpen={false}
    environment={[Window]}
    getA11yStatusMessage={[Function]}
    initialSelectedItem={null}
    isFetchingList={false}
    itemToString={[Function]}
    onChange={[Function]}
    onInputValueChange={[Function]}
    onOuterClick={[Function]}
    onSelect={[Function]}
    onStateChange={[Function]}
    onUserAction={[Function]}
    scrollIntoView={[Function]}
    selectedItemChanged={[Function]}
    stateReducer={[Function]}
    suppressRefError={false}
  >
    <div
      aria-expanded={false}
      aria-haspopup="listbox"
      aria-labelledby="downshift-0-label"
      aria-owns={null}
      role="combobox"
    >
      <div
        className="emotion-1 null"
      >
        <input
          aria-activedescendant={null}
          aria-autocomplete="list"
          aria-controls={null}
          aria-labelledby="downshift-0-label"
          autoComplete="off"
          className="emotion-0"
          disabled={false}
          id={null}
          maxLength={100}
          onBlur={[Function]}
          onChange={[Function]}
          onKeyDown={[Function]}
          placeholder=""
          value=""
        />
        <ul
          aria-labelledby="downshift-0-label"
          id="downshift-0-menu"
          role="listbox"
        />
      </div>
    </div>
  </Downshift>
</Typeahead>
`;

exports[`Typeahead renders correctly with error state 1`] = `
.emotion-1 {
  width: 100%;
  max-width: 500px;
}

.emotion-0 {
  padding: 5px;
  box-sizing: border-box;
  width: 100%;
}

.emotion-0:focus {
  outline-color: #8de2e0;
}

<Typeahead
  className={null}
  defaultHighlightedIndex={[Function]}
  defaultIsOpen={[Function]}
  disabled={false}
  environment={[Function]}
  getA11yStatusMessage={[Function]}
  getItemId={[Function]}
  highlightedIndex={[Function]}
  id={null}
  initialHighlightedIndex={[Function]}
  initialInputValue={[Function]}
  initialIsOpen={[Function]}
  initialSelectedItem={null}
  inputId={[Function]}
  inputValue={[Function]}
  isFetchingList={false}
  isOpen={[Function]}
  itemCount={[Function]}
  itemToString={[Function]}
  items={
    Array [
      "apple",
      "pear",
      "orange",
      "grape",
      "banana",
    ]
  }
  label=""
  labelId={[Function]}
  maxLength={100}
  menuId={[Function]}
  message="Something went wrong"
  minChars={0}
  onOuterClick={[Function]}
  onSelect={[Function]}
  onStateChange={[Function]}
  onUserAction={[Function]}
  placeholder=""
  scrollIntoView={[Function]}
  selectedItem={[Function]}
  selectedItemChanged={[Function]}
  suppressRefError={[Function]}
  valid={false}
>
  <Downshift
    defaultHighlightedIndex={null}
    defaultIsOpen={false}
    environment={[Window]}
    getA11yStatusMessage={[Function]}
    initialSelectedItem={null}
    isFetchingList={false}
    itemToString={[Function]}
    onChange={[Function]}
    onInputValueChange={[Function]}
    onOuterClick={[Function]}
    onSelect={[Function]}
    onStateChange={[Function]}
    onUserAction={[Function]}
    scrollIntoView={[Function]}
    selectedItemChanged={[Function]}
    stateReducer={[Function]}
    suppressRefError={false}
  >
    <div
      aria-expanded={false}
      aria-haspopup="listbox"
      aria-labelledby="downshift-2-label"
      aria-owns={null}
      role="combobox"
    >
      <div
        className="emotion-1 null"
      >
        <input
          aria-activedescendant={null}
          aria-autocomplete="list"
          aria-controls={null}
          aria-labelledby="downshift-2-label"
          autoComplete="off"
          className="emotion-0"
          disabled={false}
          id={null}
          maxLength={100}
          onBlur={[Function]}
          onChange={[Function]}
          onKeyDown={[Function]}
          placeholder=""
          value=""
        />
        <ul
          aria-labelledby="downshift-2-label"
          id="downshift-2-menu"
          role="listbox"
        />
      </div>
      <div>
        Something went wrong
      </div>
    </div>
  </Downshift>
</Typeahead>
`;

exports[`Typeahead renders correctly with items 1`] = `
.emotion-1 {
  width: 100%;
  max-width: 500px;
}

.emotion-0 {
  padding: 5px;
  box-sizing: border-box;
  width: 100%;
}

.emotion-0:focus {
  outline-color: #8de2e0;
}

<Typeahead
  className={null}
  defaultHighlightedIndex={[Function]}
  defaultIsOpen={[Function]}
  disabled={false}
  environment={[Function]}
  getA11yStatusMessage={[Function]}
  getItemId={[Function]}
  highlightedIndex={[Function]}
  id={null}
  initialHighlightedIndex={[Function]}
  initialInputValue={[Function]}
  initialIsOpen={[Function]}
  initialSelectedItem={null}
  inputId={[Function]}
  inputValue={[Function]}
  isFetchingList={false}
  isOpen={[Function]}
  itemCount={[Function]}
  itemToString={[Function]}
  items={
    Array [
      "apple",
      "pear",
      "orange",
      "grape",
      "banana",
    ]
  }
  label=""
  labelId={[Function]}
  maxLength={100}
  menuId={[Function]}
  message=""
  minChars={0}
  onOuterClick={[Function]}
  onSelect={[Function]}
  onStateChange={[Function]}
  onUserAction={[Function]}
  placeholder=""
  scrollIntoView={[Function]}
  selectedItem={[Function]}
  selectedItemChanged={[Function]}
  suppressRefError={[Function]}
  valid={true}
>
  <Downshift
    defaultHighlightedIndex={null}
    defaultIsOpen={false}
    environment={[Window]}
    getA11yStatusMessage={[Function]}
    initialSelectedItem={null}
    isFetchingList={false}
    itemToString={[Function]}
    onChange={[Function]}
    onInputValueChange={[Function]}
    onOuterClick={[Function]}
    onSelect={[Function]}
    onStateChange={[Function]}
    onUserAction={[Function]}
    scrollIntoView={[Function]}
    selectedItemChanged={[Function]}
    stateReducer={[Function]}
    suppressRefError={false}
  >
    <div
      aria-expanded={false}
      aria-haspopup="listbox"
      aria-labelledby="downshift-1-label"
      aria-owns={null}
      role="combobox"
    >
      <div
        className="emotion-1 null"
      >
        <input
          aria-activedescendant={null}
          aria-autocomplete="list"
          aria-controls={null}
          aria-labelledby="downshift-1-label"
          autoComplete="off"
          className="emotion-0"
          disabled={false}
          id={null}
          maxLength={100}
          onBlur={[Function]}
          onChange={[Function]}
          onKeyDown={[Function]}
          placeholder=""
          value=""
        />
        <ul
          aria-labelledby="downshift-1-label"
          id="downshift-1-menu"
          role="listbox"
        />
      </div>
    </div>
  </Downshift>
</Typeahead>
`;
